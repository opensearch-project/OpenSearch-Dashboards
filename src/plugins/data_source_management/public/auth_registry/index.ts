/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export {
  IAuthenticationMethodRegistry,
  AuthenticationMethod,
  AuthenticationMethodRegistry,
} from './authentication_methods_registry';
