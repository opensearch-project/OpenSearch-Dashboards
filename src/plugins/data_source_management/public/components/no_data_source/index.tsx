/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export { NoDataSource } from './no_data_source';
