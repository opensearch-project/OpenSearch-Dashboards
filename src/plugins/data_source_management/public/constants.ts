/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DirectQueryDatasourceType } from './types';

export const QUERY_RESTRICTED = 'query-restricted';
export const QUERY_ALL = 'query-all';

export const DatasourceTypeToDisplayName: { [key in DirectQueryDatasourceType]: string } = {
  PROMETHEUS: 'Prometheus',
  S3GLUE: 'Amazon S3',
};

export const PROMETHEUS_URL = 'Prometheus';
export const AMAZON_S3_URL = 'AmazonS3AWSGlue';
export const OPENSEARCH_URL = 'OpenSearch';

export const UrlToDatasourceType: { [key: string]: DirectQueryDatasourceType } = {
  [PROMETHEUS_URL]: 'PROMETHEUS',
  [AMAZON_S3_URL]: 'S3GLUE',
};
