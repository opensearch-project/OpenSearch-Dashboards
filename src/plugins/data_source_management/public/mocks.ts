/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { throwError } from 'rxjs';
import { SavedObjectsClientContract } from 'opensearch-dashboards/public';
import { AuthType } from './types';
import { coreMock } from '../../../core/public/mocks';

/* Mock Types */

export const docLinks = {
  links: {
    noDocumentation: {
      indexPatterns: {
        introduction: '',
      },
      scriptedFields: {},
    },
  },
};

const createDataSourceManagementContext = () => {
  const {
    chrome,
    application,
    savedObjects,
    uiSettings,
    notifications,
    overlays,
  } = coreMock.createStart();
  const { http } = coreMock.createSetup();

  return {
    chrome,
    application,
    savedObjects,
    uiSettings,
    notifications,
    overlays,
    http,
    docLinks,
    setBreadcrumbs: () => {},
  };
};

export const mockManagementPlugin = {
  createDataSourceManagementContext,
  docLinks,
};

/* Mock data responses - JSON*/
export const getDataSourcesResponse = {
  savedObjects: [
    {
      id: 'test',
      type: 'data-source',
      description: 'test datasource',
      title: 'test',
      get(field: string) {
        const me: any = this || {};
        return me[field];
      },
    },
    {
      id: 'test2',
      type: 'data-source',
      description: 'test datasource2',
      title: 'test',
      get(field: string) {
        const me: any = this || {};
        return me[field];
      },
    },
    {
      id: 'alpha-test',
      type: 'data-source',
      description: 'alpha test datasource',
      title: 'alpha-test',
      get(field: string) {
        const me: any = this || {};
        return me[field];
      },
    },
    {
      id: 'beta-test',
      type: 'data-source',
      description: 'beta test datasource',
      title: 'beta-test',
      get(field: string) {
        const me: any = this || {};
        return me[field];
      },
    },
  ],
};

export const existingDatasourceNamesList = [
  'test123',
  'testTest20',
  'TeSt',
  'duplicateTest',
  'dup20',
];

export const getMappedDataSources = [
  {
    id: 'test',
    description: 'test datasource',
    title: 'test',
    sort: 'test',
  },
  {
    id: 'test2',
    description: 'test datasource2',
    title: 'test',
    sort: 'test',
  },
  {
    id: 'alpha-test',
    description: 'alpha test datasource',
    title: 'alpha-test',
    sort: 'alpha-test',
  },
  {
    id: 'beta-test',
    description: 'beta test datasource',
    title: 'beta-test',
    sort: 'beta-test',
  },
];

export const mockDataSourceAttributesWithAuth = {
  id: 'test',
  title: 'create-test-ds',
  description: 'jest testing',
  endpoint: 'https://test.com',
  auth: {
    type: AuthType.UsernamePasswordType,
    credentials: {
      username: 'test123',
      password: 'test123',
    },
  },
};

export const mockDataSourceAttributesWithNoAuth = {
  id: 'test123',
  title: 'create-test-ds123',
  description: 'jest testing',
  endpoint: 'https://test.com',
  auth: {
    type: AuthType.NoAuth,
    credentials: undefined,
  },
};
export const getDataSourceByIdWithCredential = {
  attributes: {
    id: 'alpha-test',
    title: 'alpha-test',
    endpoint: 'https://test.com',
    description: 'alpha test data source',
    auth: {
      type: AuthType.UsernamePasswordType,
      credentials: {
        username: 'test123',
      },
    },
  },
};

export const getDataSourceByIdWithoutCredential = {
  attributes: {
    ...getDataSourceByIdWithCredential.attributes,
    auth: {
      type: AuthType.NoAuth,
      credentials: undefined,
    },
  },
  references: [],
};

export const mockResponseForSavedObjectsCalls = (
  savedObjectsClient: SavedObjectsClientContract,
  savedObjectsMethodName: 'get' | 'find' | 'create' | 'delete' | 'update',
  response: any
) => {
  (savedObjectsClient[savedObjectsMethodName] as jest.Mock).mockResolvedValue(response);
};

export const mockErrorResponseForSavedObjectsCalls = (
  savedObjectsClient: SavedObjectsClientContract,
  savedObjectsMethodName: 'get' | 'find' | 'create' | 'delete' | 'update'
) => {
  (savedObjectsClient[savedObjectsMethodName] as jest.Mock).mockRejectedValue(
    throwError(new Error('Error while fetching data sources'))
  );
};
