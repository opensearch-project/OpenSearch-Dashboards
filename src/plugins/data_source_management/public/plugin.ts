/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CoreSetup, CoreStart, Plugin } from '../../../core/public';

import { PLUGIN_NAME } from '../common';

import { ManagementSetup } from '../../management/public';
import { IndexPatternManagementSetup } from '../../index_pattern_management/public';
import { DataSourceColumn } from './components/data_source_column/data_source_column';

export interface DataSourceManagementSetupDependencies {
  management: ManagementSetup;
  indexPatternManagement: IndexPatternManagementSetup;
}

const DSM_APP_ID = 'dataSources';

export class DataSourceManagementPlugin
  implements Plugin<void, void, DataSourceManagementSetupDependencies> {
  public setup(
    core: CoreSetup,
    { management, indexPatternManagement }: DataSourceManagementSetupDependencies
  ) {
    const opensearchDashboardsSection = management.sections.section.opensearchDashboards;

    if (!opensearchDashboardsSection) {
      throw new Error('`opensearchDashboards` management section not found.');
    }

    const savedObjectPromise = core
      .getStartServices()
      .then(([coreStart]) => coreStart.savedObjects);
    const httpPromise = core.getStartServices().then(([coreStart]) => coreStart.http);
    const column = new DataSourceColumn(savedObjectPromise, httpPromise);
    indexPatternManagement.columns.register(column);

    opensearchDashboardsSection.registerApp({
      id: DSM_APP_ID,
      title: PLUGIN_NAME,
      order: 1,
      mount: async (params) => {
        const { mountManagementSection } = await import('./management_app');

        return mountManagementSection(core.getStartServices, params);
      },
    });
  }

  public start(core: CoreStart) {}

  public stop() {}
}
