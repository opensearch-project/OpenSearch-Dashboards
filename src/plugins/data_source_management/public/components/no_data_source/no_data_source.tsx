/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import { i18n } from '@osd/i18n';

import {
  EuiButton,
  EuiPanel,
  EuiPopover,
  EuiText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiButtonIcon,
} from '@elastic/eui';
import { ApplicationStart } from 'opensearch-dashboards/public';
import { FormattedMessage } from 'react-intl';
import { DataSourceDropDownHeader } from '../drop_down_header';
import { DSM_APP_ID } from '../../plugin';
import { EmptyIcon } from '../custom_database_icon';
import {
  ADD_COMPATIBLE_DATASOURCES_MESSAGE,
  CONNECT_DATASOURCES_MESSAGE,
  NO_COMPATIBLE_DATASOURCES_MESSAGE,
  NO_DATASOURCES_CONNECTED_MESSAGE,
} from '../constants';

interface DataSourceDropDownHeaderProps {
  incompatibleDataSourcesExist: boolean;
  application?: ApplicationStart;
}

export const NoDataSource: React.FC<DataSourceDropDownHeaderProps> = ({
  application,
  incompatibleDataSourcesExist,
}) => {
  const [showPopover, setShowPopover] = useState<boolean>(false);
  const button = (
    <EuiButtonIcon
      className="euiHeaderLink"
      data-test-subj="dataSourceEmptyMenuHeaderLink"
      aria-label={i18n.translate('dataSourceEmpty.dataSourceEmptyMenuHeaderLink', {
        defaultMessage: 'dataSourceEmptyMenuHeaderLink',
      })}
      iconType={() => <EmptyIcon />}
      size="s"
      onClick={() => setShowPopover(!showPopover)}
    />
  );

  const redirectButton = (
    <EuiButton
      data-test-subj="dataSourceEmptyStateManageDataSourceButton"
      fill={false}
      size="s"
      onClick={() =>
        application?.navigateToApp('management', {
          path: `opensearch-dashboards/${DSM_APP_ID}`,
        })
      }
    >
      <FormattedMessage
        id="dataSourcesManagement.dataSourceMenu.manageDataSource"
        defaultMessage="Manage data sources"
      />
    </EuiButton>
  );
  const text = (
    <>
      <EuiText size="s" textAlign="center">
        {
          <FormattedMessage
            id="dataSourcesManagement.dataSourceEmptyMenu.noData"
            defaultMessage={
              incompatibleDataSourcesExist
                ? NO_COMPATIBLE_DATASOURCES_MESSAGE
                : NO_DATASOURCES_CONNECTED_MESSAGE
            }
          />
        }
      </EuiText>

      <EuiText size="s" textAlign="center">
        {
          <FormattedMessage
            id="dataSourcesManagement.dataSourceEmptyMenu.connect"
            defaultMessage={
              incompatibleDataSourcesExist
                ? ADD_COMPATIBLE_DATASOURCES_MESSAGE
                : CONNECT_DATASOURCES_MESSAGE
            }
          />
        }
      </EuiText>
    </>
  );

  return (
    <EuiPopover
      id={'dataSourceEmptyStatePopover'}
      button={button}
      isOpen={showPopover}
      closePopover={() => setShowPopover(false)}
      panelPaddingSize="none"
      anchorPosition="downLeft"
      data-test-subj={'dataSourceEmptyStatePopover'}
    >
      <DataSourceDropDownHeader totalDataSourceCount={0} application={application} />
      <EuiPanel
        hasBorder={false}
        hasShadow={false}
        className="dataSourceEmptyStatePanel"
        data-test-subj="dataSourceEmptyStatePanel"
      >
        <EuiFlexGroup justifyContent={'center'}>
          <EuiFlexItem grow={false}>{text}</EuiFlexItem>
        </EuiFlexGroup>
        <EuiFlexGroup justifyContent={'center'}>
          <EuiFlexItem grow={false}>{redirectButton}</EuiFlexItem>
        </EuiFlexGroup>
      </EuiPanel>
    </EuiPopover>
  );
};
