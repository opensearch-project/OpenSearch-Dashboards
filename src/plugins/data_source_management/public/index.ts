/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DataSourceManagementPlugin } from './plugin';

// This exports static code and TypeScript types,
// as well as, OpenSearch Dashboards Platform `plugin()` initializer.
export function plugin() {
  return new DataSourceManagementPlugin();
}
export { DataSourceManagementPluginStart } from './types';
