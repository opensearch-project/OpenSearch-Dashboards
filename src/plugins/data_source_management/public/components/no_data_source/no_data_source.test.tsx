/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ShallowWrapper, shallow } from 'enzyme';
import React from 'react';
import { render } from '@testing-library/react';
import { NoDataSource } from './no_data_source';
import { coreMock } from '../../../../../core/public/mocks';
import { DSM_APP_ID } from '../../plugin';

describe('NoDataSource', () => {
  let component: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
  const totalDataSourceCount = 0;
  const nextTick = () => new Promise((res) => process.nextTick(res));

  it('should render correctly with the provided totalDataSourceCount', () => {
    const wrapper = shallow(
      <NoDataSource totalDataSourceCount={totalDataSourceCount} hasIncompatibleDatasource={false} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should display popover when click "No data sources" button', async () => {
    const applicationMock = coreMock.createStart().application;
    const container = render(
      <NoDataSource
        totalDataSourceCount={totalDataSourceCount}
        application={applicationMock}
        hasIncompatibleDatasource={false}
      />
    );

    await nextTick();

    const button = await container.findByTestId('dataSourceEmptyMenuHeaderLink');
    button.click();

    expect(container.getByTestId('dataSourceEmptyStatePopover')).toBeVisible();
  });

  it('should call application.navigateToApp when the "Manage" link is clicked', async () => {
    const applicationMock = coreMock.createStart().application;
    const navigateToAppMock = applicationMock.navigateToApp;

    const container = render(
      <NoDataSource
        totalDataSourceCount={totalDataSourceCount}
        application={applicationMock}
        hasIncompatibleDatasource={false}
      />
    );

    await nextTick();

    const button = await container.findByTestId('dataSourceEmptyMenuHeaderLink');
    button.click();
    const redirectButton = await container.findByTestId(
      'dataSourceEmptyStateManageDataSourceButton'
    );
    redirectButton.click();
    expect(navigateToAppMock).toHaveBeenCalledWith('management', {
      path: `opensearch-dashboards/${DSM_APP_ID}`,
    });
  });

  it.each([false, true])(
    'should render normally when incompatibleDataSourcesExist is %b',
    (incompatibleDataSourcesExist) => {
      component = shallow(
        <NoDataSource
          totalDataSourceCount={0}
          hasIncompatibleDatasource={incompatibleDataSourcesExist}
        />
      );
      expect(component).toMatchSnapshot();
    }
  );
});
