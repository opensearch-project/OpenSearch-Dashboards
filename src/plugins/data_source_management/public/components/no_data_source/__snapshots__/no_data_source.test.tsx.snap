// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoDataSource should render correctly with the provided totalDataSourceCount 1`] = `
<EuiPopover
  anchorPosition="downLeft"
  button={
    <EuiButtonIcon
      aria-label="dataSourceEmptyMenuHeaderLink"
      className="euiHeaderLink"
      data-test-subj="dataSourceEmptyMenuHeaderLink"
      iconType={[Function]}
      onClick={[Function]}
      size="s"
    />
  }
  closePopover={[Function]}
  data-test-subj="dataSourceEmptyStatePopover"
  display="inlineBlock"
  hasArrow={true}
  id="dataSourceEmptyStatePopover"
  isOpen={false}
  ownFocus={true}
  panelPaddingSize="none"
>
  <DataSourceDropDownHeader
    totalDataSourceCount={0}
  />
  <EuiPanel
    className="dataSourceEmptyStatePanel"
    data-test-subj="dataSourceEmptyStatePanel"
    hasBorder={false}
    hasShadow={false}
  >
    <EuiFlexGroup
      justifyContent="center"
    >
      <EuiFlexItem
        grow={false}
      >
        <EuiText
          size="s"
          textAlign="center"
        >
          <FormattedMessage
            defaultMessage="No data sources connected yet."
            id="dataSourcesManagement.dataSourceEmptyMenu.noData"
            values={Object {}}
          />
        </EuiText>
        <EuiText
          size="s"
          textAlign="center"
        >
          <FormattedMessage
            defaultMessage="Connect your data sources to get started."
            id="dataSourcesManagement.dataSourceEmptyMenu.connect"
            values={Object {}}
          />
        </EuiText>
      </EuiFlexItem>
    </EuiFlexGroup>
    <EuiFlexGroup
      justifyContent="center"
    >
      <EuiFlexItem
        grow={false}
      >
        <EuiButton
          data-test-subj="dataSourceEmptyStateManageDataSourceButton"
          fill={false}
          onClick={[Function]}
          size="s"
        >
          <FormattedMessage
            defaultMessage="Manage data sources"
            id="dataSourcesManagement.dataSourceMenu.manageDataSource"
            values={Object {}}
          />
        </EuiButton>
      </EuiFlexItem>
    </EuiFlexGroup>
  </EuiPanel>
</EuiPopover>
`;

exports[`NoDataSource should render normally 1`] = `
<EuiPopover
  anchorPosition="downLeft"
  button={
    <EuiButtonIcon
      aria-label="dataSourceEmptyMenuHeaderLink"
      className="euiHeaderLink"
      data-test-subj="dataSourceEmptyMenuHeaderLink"
      iconType={[Function]}
      onClick={[Function]}
      size="s"
    />
  }
  closePopover={[Function]}
  data-test-subj="dataSourceEmptyStatePopover"
  display="inlineBlock"
  hasArrow={true}
  id="dataSourceEmptyStatePopover"
  isOpen={false}
  ownFocus={true}
  panelPaddingSize="none"
>
  <DataSourceDropDownHeader
    totalDataSourceCount={0}
  />
  <EuiPanel
    className="dataSourceEmptyStatePanel"
    data-test-subj="dataSourceEmptyStatePanel"
    hasBorder={false}
    hasShadow={false}
  >
    <EuiFlexGroup
      justifyContent="center"
    >
      <EuiFlexItem
        grow={false}
      >
        <EuiText
          size="s"
          textAlign="center"
        >
          <FormattedMessage
            defaultMessage="No data sources connected yet."
            id="dataSourcesManagement.dataSourceEmptyMenu.noData"
            values={Object {}}
          />
        </EuiText>
        <EuiText
          size="s"
          textAlign="center"
        >
          <FormattedMessage
            defaultMessage="Connect your data sources to get started."
            id="dataSourcesManagement.dataSourceEmptyMenu.connect"
            values={Object {}}
          />
        </EuiText>
      </EuiFlexItem>
    </EuiFlexGroup>
    <EuiFlexGroup
      justifyContent="center"
    >
      <EuiFlexItem
        grow={false}
      >
        <EuiButton
          data-test-subj="dataSourceEmptyStateManageDataSourceButton"
          fill={false}
          onClick={[Function]}
          size="s"
        >
          <FormattedMessage
            defaultMessage="Manage data sources"
            id="dataSourcesManagement.dataSourceMenu.manageDataSource"
            values={Object {}}
          />
        </EuiButton>
      </EuiFlexItem>
    </EuiFlexGroup>
  </EuiPanel>
</EuiPopover>
`;
